### PR TITLE
Fix ConstantPad2dStaticModule test.

### PR DESCRIPTION
**Pull request by gprateek93**
This commit fixes the `ConstantPad2dStaticModule` test case by adding
the lowering of `aten.pad` operation. Previously the test case
mapped to `aten.constant_pad_nd` operation.
The `aten.pad` now decomposes into `aten.constant_pad_nd` operation.

Signed-Off-By: Prateek Gupta <prateek@nod-labs.com>

### CODE REVIEW

**Comment by gprateek93** [comment]
> Please add the decompose_complex_ops test.

Thanks! Added the test.

**Comment by pashu123** [comment]
> > Please add the decompose_complex_ops test.
> 
> Thanks! Added the test.

I think you also need to add an e2e test of pad op.

**Comment by gprateek93** [comment]
> > > Please add the decompose_complex_ops test.
> > 
> > 
> > Thanks! Added the test.
> 
> I think you also need to add an e2e test of pad op.

It is already added. That is what this PR fixes. Previously this test case lowered to `aten.constant_pad_nd` but due to latest pytorch version, now it lowers to `aten.pad` operation.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -1646,6 +1646,25 @@ class DecomposeAtenNewEmptyOp : public OpRewritePattern<AtenNewEmptyOp> {
...
 
+namespace {
+// Decompose `aten.pad` op into `aten.constant_pad_nd` op.
+class DecomposeAtenPadOp : public OpRewritePattern<AtenPadOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenPadOp op,
+                                PatternRewriter &rewriter) const override {
+
+    Value value = op.value();
+    if (!value)
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1657]: should check for NoneType here.

Also, need to guard for OptionalType. This is because we can't really handle optional type but only NoneType or FloatType.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]
Shall we change the e2e tests to use the `torch.ops.aten.` operators directly? We could also test the padding value None case.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
Comment (gprateek93) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Added.

**Comment by silvasean** [comment]
Can you summarize why this test suddenly started failing? Was it an upstream change?

**Comment by pashu123** [comment]
> Can you summarize why this test suddenly started failing? Was it an upstream change?

I think it's failing because of `"GeFloatModule_basic"` which you have already reverted. Rebasing the patch would fix the issue.

**Comment by silvasean** [comment]
I mean why did ConstantPad2dStaticModule start failing in the first place (why did we have to fix it?)

**Comment by gprateek93** [comment]
> I mean why did ConstantPad2dStaticModule start failing in the first place (why did we have to fix it?)

Yes it started failing due to upstream change. Previously the `torch.nn.ConstantPad2d` op matched to `aten.constant_pad_nd` operation, but after the introduction of `aten.pad` recently, the op matched to it.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -1664,6 +1665,25 @@ class DecomposeAtenIndexPutHackedTwinOp
...
 } // namespace
 
+namespace {
+// Decompose `aten.pad` op into `aten.constant_pad_nd` op.
+class DecomposeAtenPadOp : public OpRewritePattern<AtenPadOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenPadOp op,
+                                PatternRewriter &rewriter) const override {
+
+    Value value = op.value();
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1675]: need to guard for OptionalType as well. If it's OptionalType need to report error.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (gprateek93) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1675]: Added.
